### PR TITLE
Start ladspa plugins from script + service improvements [sam dev 3/3]

### PR DESCRIPTION
**Pull request by sandreae**
- Use ladspa plugins for all connections above 2 clients
- Create method for connecting a list of jacktrip clients to a list of ladspa plugins
- Automated starting/stopping of ladspa plugins.
- jackd restarts when a depended pypatcher service fails.
- Improvements to service scripts including teal-time permissions 
- also includes some fixes for the previous 2 branches in this chain

### CODE REVIEW

**Comment by madwort** [comment]
Wow, lots of great stuff in here - would it be possible to split it out into multiple pull requests though...? The changes to the service scripts look fine, the refactor of  `jack_client_patching` looks good (I think), but I think I have some things to look at in more depth in the ladspa scripts, but there's so much going on at once! Give me a shout if you need any pointers - it looks like everything in is sensibly labelled commits so if you create some fresh branches off `master` and `git cherry-pick` the relevant commits you may be able to make several branches (one for each feature) fairly easily? Would make it waaaay more easy to review.

**Comment by sandreae** [comment]
Yeh, knew I should have done that 😅

Actually I worked on 3 seperate branches for these features and then merged them all into dev when they were tested, so I can do some PRs from each of those. 

Watch this space...

File: jacktrip_pypatcher.py [hunk 1]
```diff
@@ -4,14 +4,16 @@
 import stereo_recording
 import random
 import jack_client_patching as p
+from itertools import product
```
Comment (madwort) [inline jacktrip_pypatcher.py:7]: is this being used? seems like not?

File: ladspa_plugins.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        [
+            jackspa_path,
+            "-j",
+            "ladspa-left-30",
+            "-i",
+            "0:0:0:-0.30:0:0",
+            "/usr/lib/ladspa/inv_input.so",
+            "3301",
+        ]
+    )
```
Comment (madwort) [inline ladspa_plugins.py:39]: probably worth doing a helper function for this, like `start_plugin('ladspa-left-30',-0.3)` to reduce duplication in this file?

File: jacktrip_pypatcher.py [hunk 3]
```diff
@@ -67,27 +77,49 @@ def get_darkice_port(jackClient, dry_run, darkice_prefix):
...
+    ladspa_left_15 = "ladspa-left-15"
+    ladspa_left_30 = "ladspa-left-30"
+    ladspa_left_45 = "ladspa-left-45"
+    ladspa_left_60 = "ladspa-left-60"
+    ladspa_left_75 = "ladspa-left-75"
+    ladspa_right_15 = "ladspa-right-15"
+    ladspa_right_30 = "ladspa-right-30"
+    ladspa_right_45 = "ladspa-right-45"
+    ladspa_right_60 = "ladspa-right-60"
+    ladspa_right_75 = "ladspa-right-75"
```
Comment (madwort) [inline jacktrip_pypatcher.py:100]: hmm, this list also appears implicitly duplicated in the ladspa file - would be really good to merge them somehow. I'm not quite sure how this would look - I think probably a couple of static arrays with lists of pan positions, but I'm not sure whether the arrays live in this file or the ladspa file. Also, maybe it's just a list of positions, and a helper function that generates the names automatically, e.g. `[0, 0.15, 0.3, ...]` and `ladspa_port_name(0.15)` returns `'ladspa-right-15'`? would be good to make this more generic, and make the choice of panning more flexible/configurable.

File: ladspa_plugins.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    subprocess.Popen(
+        [
+            jackspa_path,
+            "-j",
+            "ladspa-right-30",
```
Comment (madwort) [inline ladspa_plugins.py:44]: this string has to match up with the value in the main file (this is the duplication that I mean).

File: jacktrip_pypatcher.py [hunk 5]
```diff
@@ -67,27 +77,49 @@ def get_darkice_port(jackClient, dry_run, darkice_prefix):
...
 
-    if len(jacktrip_clients) > 3:
-        verify_ladspa_plugins(jackClient)
+    if len(jacktrip_clients) >= 2 and len(all_ladspa_ports) < 24:
+        ladspa_plugins.start_plugins_2(jackClient)
+
+    if len(jacktrip_clients) >= 5 and len(all_ladspa_ports) < 32:
+        ladspa_plugins.start_plugins_5(jackClient)
+
+    if len(jacktrip_clients) >= 10 and len(all_ladspa_ports) < 40:
```
Comment (madwort) [inline jacktrip_pypatcher.py:119]: nifty idea to count the ports, but might be unreliable if multiple clients join in between runs? feels like there might be a way to make this a bit more robust.

File: jacktrip_pypatcher.py (hunk 3 above)
Comment (madwort) [reply jacktrip_pypatcher.py:100]: I guess what I'm maybe saying is, `jacktrip_pypatcher` has an array like
`ladspa_positions_2_to_5_clients = [0.3, -0.3, 0.45, -0.45, 0.6, -0.6]`
and instead of calling `ladspa.start_ladspa_2(..)` calls `ladspa.start(ladspa_positions_2_to_5_clients)`? and both of them use `ladspa.get_port_name(ladspa_positions_2_to_5_clients[n])` when they need the name?

Comment (madwort) [reply jacktrip_pypatcher.py:100]: also with an eye on how to go from 11 clients to 111 clients ;)

Comment (madwort) [reply jacktrip_pypatcher.py:100]: but also - what if you want to tweak the pan positions later

File: services/jacktrip.service [hunk 6]
```diff
@@ -4,7 +4,11 @@ Requires=jackd.service
 After=jackd.service
 
 [Service]
-ExecStart=/usr/bin/jacktrip -S -p0 --nojackportsconnect -z
+LimitRTPRIO=infinity
+LimitMEMLOCK=infinity
+LimitRTTIME=infinity
+EnvironmentFile=/etc/jacktrip_pypatcher/jacktrip.conf
+ExecStart=/usr/bin/jacktrip -S -p0 --nojackportsconnect -z -q $Q
```
Comment (madwort) [inline services/jacktrip.service:11]: service files remain good - if they were in their own PR would merge them straight away

File: jacktrip_pypatcher.py (hunk 5 above)
Comment (madwort) [reply jacktrip_pypatcher.py:119]: if we had an array of pan positions available, we could actually iterate through them & check individually that the specific ones we need are all running? and only start the ones that aren't running?

File: jacktrip_pypatcher.py [hunk 7]
```diff
@@ -117,312 +149,213 @@ def autopatch(jackClient, dry_run, jacktrip_clients, jacktrip_clients_stereo):
...
+
+        all_hold_music_ports = jackClient.get_ports(hold_music_port + ".*")
+        if len(all_hold_music_ports) == 0:
+            # this is a crude fix because mpg123 wasn't always starting
+            # expect we won't need this with the server upgrade as audio
+            # processing is much less error prone
+            count = 0
+            while count < 3:
+                lounge_music.start_the_music(jackClient, hold_music_port)
+                count += 1
```
Comment (madwort) [inline jacktrip_pypatcher.py:161]: should this have a `sleep` too?

File: jacktrip_pypatcher.py [hunk 8]
```diff
@@ -117,312 +149,213 @@ def autopatch(jackClient, dry_run, jacktrip_clients, jacktrip_clients_stereo):
...
+        jcp.connect_to_ladspa(jacktrip_clients[2], ladspa_right_45)
+
+        jcp.connect_from_ladspa(ladspa_left_45, jacktrip_clients[1])
+        jcp.connect_from_ladspa(ladspa_right_45, jacktrip_clients[1])
+
+        jcp.connect_to_ladspa(jacktrip_clients[0], ladspa_left_60)
+        jcp.connect_to_ladspa(jacktrip_clients[1], ladspa_right_60)
+
+        jcp.connect_from_ladspa(ladspa_left_60, jacktrip_clients[2])
+        jcp.connect_from_ladspa(ladspa_right_60, jacktrip_clients[2])
```
Comment (madwort) [inline jacktrip_pypatcher.py:204]: ...haven't got my head around this block yet...

File: jacktrip_pypatcher.py (hunk 3 above)
Comment (sandreae) [reply jacktrip_pypatcher.py:100]: Yeh, It didn't feel good to duplicate these lists of ladspa_ports everywhere.... Thanks for the suggestion, that really makes sense. Definitely good to have flexibility with the panning positions as I'm sure we'll want to fine tune things as we do more sessions. I'll have a go at putting something together :+1:

File: ladspa_plugins.py (hunk 4 above)
Comment (sandreae) [reply ladspa_plugins.py:44]: gotcha

File: jacktrip_pypatcher.py (hunk 5 above)
Comment (sandreae) [reply jacktrip_pypatcher.py:119]: Yeh, good point, will look into this along with the above changes.

File: jacktrip_pypatcher.py (hunk 8 above)
Comment (sandreae) [reply jacktrip_pypatcher.py:204]: Yeh, this is quite weird.... It came up as a result of not routing clients directly to each other and instead inserting panning plugins in-between (done so that stereo clients aren't patched down to mono hard-left or hard-right). However taking the same approach as sessions with more clients would mean that each clients send mix would be unbalanced. Like this:

```
# sends to ladspa
jcp.connect_to_ladspa(jacktrip_clients[0], ladspa_left_45)
jcp.connect_to_ladspa(jacktrip_clients[1], ladspa_centre)
jcp.connect_to_ladspa(jacktrip_clients[2], ladspa_right_45)

# jacktrip_clients[0] mix <--- bad mix
jcp.connect_from_ladspa(ladspa_centre, jacktrip_clients[0])
jcp.connect_from_ladspa(ladspa_right_45, jacktrip_clients[0])

# jacktrip_clients[1] mix <--- ok mix
jcp.connect_from_ladspa(ladspa_left_45, jacktrip_clients[1])
jcp.connect_from_ladspa(ladspa_right_45, jacktrip_clients[1])

# jacktrip_clients[2] mix <--- bad mix
jcp.connect_from_ladspa(ladspa_left_45, jacktrip_clients[2])
jcp.connect_from_ladspa(ladspa_centre, jacktrip_clients[2])
```

As a solution I used different left-right panning positions for each client receive, so although each client still has a slightly different send mix, at least they are all left-right balanced. I may well have missed an obvious solution, but this is all I could come up with...

File: jacktrip_pypatcher.py (hunk 3 above)
Comment (sandreae) [reply jacktrip_pypatcher.py]: Midway through having a go here: https://github.com/noiseorchestra/jacktrip_pypatcher/pull/25/commits/ff0c6fe256d8360972e5de6b0210effd410fefe3

It's slimmed `ladspa_plugins` right down, very tidy. But it's a bit more complex in `jacktrip_pypatcher` because each number of clients requires a specific set of panning positions, and these don't really increment predictably. It could be changed to make the patching slightly more arbitrary, which would scale well, but then you'd lose the fine tuning which I think we want atm.

File: jacktrip_pypatcher.py [hunk 9]
```diff
@@ -77,6 +73,25 @@ def get_darkice_port(jackClient, dry_run, darkice_prefix):
...
+        return "ladspa-right-" + str(int(position*100))
+
+
+def get_ladspa_port_names(panning_positions):
+    """Returns a list of ladspa port names"""
+    # Currently not using this, just an example for now.
+    port_names = []
+    for position in panning_positions:
+        port_names.append(get_ladspa_port_name(position))
+    return port_names
```
Comment (madwort) [inline jacktrip_pypatcher.py:92]: this is classic imperative loop style, but we could also do functional programming style something like:
```
def get_ladspa_port_names(panning_positions):
    """Returns a list of ladspa port names"""
    return map(get_ladspa_port_name, panning_positions)
```
and if we do that, then maybe we don't really need this helper function anymore, we just inline the map when we need to do this?

File: test_jacktrip_pypatcher.py [hunk 10]
```diff
@@ -1,6 +1,30 @@
...
+    cmd = ["/home/sam/ng-jackspa/jackspa-cli",
+           "-j",
+           "ladspa-right-30",
+           "-i",
+           "0:0:0:0.3:0:0",
+           "/usr/lib/ladspa/inv_input.so",
+           "3301",
+           ]
+
+    assert ladspa_plugins.generate_subprocess_cmd(position) == cmd
```
Comment (madwort) [inline test_jacktrip_pypatcher.py:16]: nice! tests! 🎉

File: test_jacktrip_pypatcher.py [hunk 11]
```diff
@@ -1,6 +1,30 @@
...
+def test_get_ladspa_ports():
+
+    panning_positions = [0, -0.3, 0.3, -0.6, 0.6]
+    port_names = ["ladspa-centre",
+                  "ladspa-left-30",
+                  "ladspa-right-30",
+                  "ladspa-left-60",
+                  "ladspa-right-60"]
+
+    assert jacktrip_pypatcher.get_ladspa_port_names(panning_positions) == port_names
```
Comment (madwort) [inline test_jacktrip_pypatcher.py:27]: this could be 
```
    assert map(jacktrip_pypatcher.get_ladspa_port_names,panning_positions) == port_names
```

File: jacktrip_pypatcher.py [hunk 12]
```diff
@@ -77,6 +73,25 @@ def get_darkice_port(jackClient, dry_run, darkice_prefix):
     return darkice_ports[0]
 
 
+def get_ladspa_port_name(position):
```
Comment (madwort) [inline jacktrip_pypatcher.py:76]: I feel like this function should be in `ladspa_plugin.py` rather than here - wdyt?

File: jacktrip_pypatcher.py (hunk 8 above)
Comment (madwort) [reply jacktrip_pypatcher.py]: ah I see - it feels a bit unsatisfactory to have three such different mixes for the different users, feel like there might be a better way but haven't had time to think it through yet...

**Comment by sandreae** [comment]
Another thought about the ladspa port list, how about this:

`ladspa_plugins = [0, -0.15, 0.15, -0.3, 0.3, -0.45, 0.45, -0.6, 0.6, -0.75, 0.75]`

then using it like:

```
if len(jacktrip_clients) >= 2 and len(all_ladspa_ports) < 24:
    ladspa.start_plugins(jackClient, ladspa_plugins[3:9])
```

and 

```
if len(jacktrip_clients) == 4:
     ladspa_ports = ladspa_plugins[3:7]
     connect_all(jcp, jacktrip_clients, ladspa_ports)

    print("-- darkice --")
    for ladspa_port in ladspa_ports:
        jcp.connect_darkice_from_ladspa(ladspa_port, darkice_port)
```

It's slightly less readable at first, but when you get used to `0 == center`, `negative == left`, `positive == right` then it's ok really. Cuts out a load of repetition and scales better.
`

File: jacktrip_pypatcher.py (hunk 9 above)
Comment (sandreae) [reply jacktrip_pypatcher.py:92]: yep, agreed.

File: jacktrip_pypatcher.py (hunk 7 above)
Comment (sandreae) [reply jacktrip_pypatcher.py:186]: done: https://github.com/noiseorchestra/jacktrip_pypatcher/pull/25/commits/19b53946a0318217545435b58748e2eeb878d898

**Comment by sandreae** [comment]
With a helper function it could then be slimmed down to:

```
if len(jacktrip_clients) == 4:
    patch_session(jcp, jacktrip_clients, panning_positions[3:7], darkice_port)
if len(jacktrip_clients) == 5:
    patch_session(jcp, jacktrip_clients, [panning_positions[0]] + panning_positions[3:7], darkice_port)
```
etc...

**Comment by madwort** [comment]
yeah, that's neat, but what worries me is about - what if the future programmer makes a typo in the segments of the panning positions, like:

```python
if len(jacktrip_clients) >= 2 and len(all_ladspa_ports) < 24:
    ladspa.start_plugins(jackClient, map(get_ladspa_port_name, panning_positions[3:9]))
```

with 

```python
if len(jacktrip_clients) == 4:
     ladspa_ports = map(get_ladspa_port_name, panning_positions[2:7]) // should be [3:7]
     connect_all(jcp, jacktrip_clients, ladspa_ports)
) 
```

What about if we did:

```python
panning_positions_2_to_5 = panning_positions[3:9]

if len(jacktrip_clients) >= 2 and len(all_ladspa_ports) < 24:
    ladspa.start_plugins(jackClient, map(get_ladspa_port_name, panning_positions_2_to_5))
```

with 

```python
if len(jacktrip_clients) == 4:
     ladspa_ports = map(get_ladspa_port_name, panning_positions_2_to_5[0:5]) // equivalent to panning_positions[3:7]
     connect_all(jcp, jacktrip_clients, ladspa_ports)
) 
```

on the downside, you have to think about more about it, but on the upside, there's no way that you can attempt to use a panning position that doesn't exist. (we're talking about safe python arrays here, but this is a bit like doing pointers in languages like C, in which case if you bust out of your array you get a memory protection error & segfault...)

**Comment by madwort** [comment]
but actually I'm starting to wonder about a function something like

```python
def ladspa.panning_positions(number_of_clients):
  if number_of_clients == 2:
    return panning_positions[3:7] // or whatever
```
and then we could:
```python
panning_positions = ladspa.panning_positions(2)
ladspa.start_plugins(jackClient, map(get_ladspa_port_name, panning_positions))
...
if len(jacktrip_clients) == 4:
     ladspa_ports = map(get_ladspa_port_name, panning_positions) 
     connect_all(jcp, jacktrip_clients, ladspa_ports)
) 
```
if we modified `get_ladspa_port_name` to verify whether the port existed & start the plugin if it doesn't?
Then we can put all of this stuff into one neat helper function in `ladspa` with no side effects and write some easy unit tests against it (because we pass in an int & get an array in return).

**Comment by sandreae** [comment]
Ok, kind of a flood of changes here, all stuff we discussed above. Hopefully it's clear from the commits what is going on....

**Comment by sandreae** [comment]
There's still this the around connecting 3 clients..... would be good if we find a neater solution for this.

File: jacktrip_pypatcher.py (hunk 12 above)
Comment (sandreae) [reply jacktrip_pypatcher.py]: done: https://github.com/noiseorchestra/jacktrip_pypatcher/pull/25/commits/4343123e9a8dd6fc5e497654f84bfcf414114675

File: ladspa_plugins.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  if number_of_clients == 7:
+    return all_panning_positions[0:3] + all_panning_positions[5:9]
+  if number_of_clients == 8:
+    return all_panning_positions[1:9]
+  if number_of_clients == 9:
+    return all_panning_positions[0:9]
+  if number_of_clients == 10:
+    return all_panning_positions[1:11]
+  if number_of_clients == 11:
+    return all_panning_positions[0:11]
```
Comment (madwort) [inline ladspa_plugins.py:25]: this could totally be a switch:
```
switch (number_of_clients):
  case 4:
    return all_panning_positions[3:7]
  ...etc...
```
should have a `default` case too that just throws some kind of error?

Comment (sandreae) [reply ladspa_plugins.py:25]: Hell yeah, love a good switch.

File: jacktrip_pypatcher.py [hunk 14]
```diff
@@ -152,51 +152,46 @@ def autopatch(jackClient, dry_run, jacktrip_clients):
...
-        ladspa_port_4 = ladspa.get_port(jackClient, all_panning_positions[4], all_ladspa_ports)
-        ladspa_port_5 = ladspa.get_port(jackClient, all_panning_positions[5], all_ladspa_ports)
-        ladspa_port_6 = ladspa.get_port(jackClient, all_panning_positions[6], all_ladspa_ports)
-        ladspa_port_7 = ladspa.get_port(jackClient, all_panning_positions[7], all_ladspa_ports)
-        ladspa_port_8 = ladspa.get_port(jackClient, all_panning_positions[8], all_ladspa_ports)
+        # if we like this method we can add the positions to all_panning_positions
+        ladspa_mid_left_1 = ladspa.get_port(jackClient, all_panning_positions[5], all_ladspa_ports)
+        ladspa_mid_left_2 = ladspa.get_port(jackClient, -0.46, all_ladspa_ports)
+        ladspa_mid_right_1 = ladspa.get_port(jackClient, all_panning_positions[6], all_ladspa_ports)
+        ladspa_mid_right_2 = ladspa.get_port(jackClient, 0.46, all_ladspa_ports)
```
Comment (sandreae) [inline jacktrip_pypatcher.py:159]: Doing it like this at least means we are only starting 4 ladspa at this point. However 2 of them aren't reused later... We *could* get a center ladspa involved at this point then 2 & 3 clients could also be passed straight into our patch_session function which would be super neat.

Comment (sandreae) [reply jacktrip_pypatcher.py:159]: Also, I'm now wondering if this...

`    all_panning_positions = [0, -0.15, 0.15, -0.3, 0.3, -0.45,
                             0.45, -0.6, 0.6, -0.75, 0.75]`

....should actually be a dict. The idea of adding a couple of extra positions and then editing our `get_panning_positions` is a bit daunting (and error prone). A dict might be a bit more faffy to unpack, but would at least be fairly explicit.

Comment (madwort) [reply jacktrip_pypatcher.py:159]: hmmm, I'm starting to think that I'm basically happy with the way the generic case (4+ clients) is working, so maybe we should test & merge, and then improve the way the edge cases (2/3 clients) work in follow-up PRs? This has come a long way in one PR!

Comment (madwort) [reply jacktrip_pypatcher.py:159]: (I'm finding it's getting to see what's what in this PR - e.g. if we merged this, and then you did a PR that converted to a dict structure, it would be much easier to see the idea, as the PR would be comparing with dict to without dict, whereas now I can only compare against the code that's before all of these huge changes)

Comment (madwort) [reply jacktrip_pypatcher.py:159]: have you been testing with multiple clients as you go along? are you confident that it's all working as intended?

Comment (sandreae) [reply jacktrip_pypatcher.py:159]: > So maybe we should test & merge, and then improve the way the edge cases (2/3 clients) work in follow-up PRs?

Yeh, I can't even remember the beginning of the PR anymore!! (was it pre-lockdown :confused: ??)

>  you been testing with multiple clients as you go along?

I've tested with multiple clients, caught a couple of typo bugs this morning, apart from that it seems to be working as expected.

> are you confident that it's all working as intended?

Urm... how confident do I need to be ;-p ? Recon I'm confident enough to merge. Our next server session is on Tuesday so I can do more checks on Monday to be sure.

Comment (madwort) [reply jacktrip_pypatcher.py:159]: cool - for one of the group sessions earlier in the year, I made some big changes the night before, then there was a typo in the code at the start of the session - frantically fixed it to get the session back on track, good to verify there's none of them!

Comment (sandreae) [reply jacktrip_pypatcher.py:159]: Ok, done some more tests with up to 8 clients, seems to be working fine. There is about half a second silence when a new client joins and the script runs, can't remember if it was this much before... But this can be sorted later if it's worse than it needs to be.

Thanks for your guidance on this epic PR!!! Gunna merge now.

Comment (madwort) [reply jacktrip_pypatcher.py:159]: hmmm, yeah, it's been in the back of my mind to improve this - at the moment the script (in general) disconnects everything, then reconnects everything - would be much better if it could only reconnect the things that need reconnecting. One thought is - maybe at the moment we do this:
* disconnect
* start ladspa
* connect

and maybe there would be less silence if we did:
* start ladspa
* disconnect
* connect

I'll make an issue ticket & assign it to you...?

Comment (sandreae) [reply jacktrip_pypatcher.py:159]: Yeh, I was thinking the same. However this only makes a difference the first time we cross each client number threshold, right? And I think there is still a good bit of silence between disconnect and connect everytime someone joins / leaves. Anyway, will follow up on the issue! Thanks :+1:
